File: 3367-maximize-sum-of-weights-after-edge-removals/3367-maximize-sum-of-weights-after-edge-removals.java
class Solution {
    static class Edge {
        int neighbour;
        int weight;

        Edge(int neighbour, int weight) {
            this.neighbour = neighbour;
            this.weight = weight;
        }
    }

    private long[][] dp;
    private List<List<Edge>> gr;
    private int k;

    public long maximizeSumOfWeights(int[][] edges, int k) {
        int n = edges.length + 1;
        this.k = k;
        gr = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            gr.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            gr.get(edge[0]).add(new Edge(edge[1], edge[2]));
            gr.get(edge[1]).add(new Edge(edge[0], edge[2]));
        }
        dp = new long[n][2];
        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], -1L);
        }
        return solve(0, 0, -1);
    }

    private long solve(int node, int isParEdgeRemoved, int parent) {
        if (dp[node][isParEdgeRemoved] != -1) return dp[node][isParEdgeRemoved];

        int edgesToRemove = Math.max(0, gr.get(node).size() - k - isParEdgeRemoved);
        long answer = 0;
        List<long[]> candidates = new ArrayList<>();
        for (Edge edge : gr.get(node)) {
            int ne = edge.neighbour, wt = edge.weight;
            if (ne != parent) {
                long dontRemove = solve(ne, 0, node) + wt;
                long remove = solve(ne, 1, node);
                candidates.add(new long[]{dontRemove, remove});
            }
        }

        candidates.sort((p1, p2) -> Long.compare(p1[0] - p1[1], p2[0] - p2[1]));
        for (int i = 0; i < candidates.size(); i++) {
            long dontRemove = candidates.get(i)[0];
            long remove = candidates.get(i)[1];
            if (i < edgesToRemove) {
                answer += remove;
            } else {
                answer += Math.max(dontRemove, remove);
            }
        }
        return dp[node][isParEdgeRemoved] = answer;
    }
}

File: 3367-maximize-sum-of-weights-after-edge-removals/3367-maximize-sum-of-weights-after-edge-removals.cpp

class Solution {
public:
    vector<vector<pair<int,int>>> gr;  // {neighbour, weight}
    vector<vector<long long>> dp;
    int k;

    long long maximizeSumOfWeights(vector<vector<int>>& edges, int k) {
    //   Number of nodes=Number of edges+1
      
        int n = edges.size() + 1;
        this->k = k;
        gr.assign(n, {});
        
        for (auto &e : edges) {
            int u = e[0], v = e[1], w = e[2];
            gr[u].push_back({v, w});
            gr[v].push_back({u, w});
        }
        
        dp.assign(n, vector<long long>(2, -1));
        return solve(0, 0, -1);
    }

    long long solve(int node, int isParEdgeRemoved, int parent) {
        if (dp[node][isParEdgeRemoved] != -1)
            return dp[node][isParEdgeRemoved];

        int degree = gr[node].size();
        int edgesToRemove = max(0, degree - k - isParEdgeRemoved);
        long long ans = 0;

        vector<pair<long long,long long>> candidates;
        for (auto &[ne, wt] : gr[node]) {
            if (ne == parent) continue;

            long long dontRemove = solve(ne, 0, node) + wt;
            long long remove = solve(ne, 1, node);
            candidates.push_back({dontRemove, remove});
        }

        // Sort by "loss" if we remove the edge
        sort(candidates.begin(), candidates.end(),
             [](auto &a, auto &b) {
                 return (a.first - a.second) < (b.first - b.second);
             });

        for (int i = 0; i < (int)candidates.size(); i++) {
            long long dontRemove = candidates[i].first;
            long long remove = candidates[i].second;
            if (i < edgesToRemove)
                ans += remove;
            else
                ans += max(dontRemove, remove);
        }

        return dp[node][isParEdgeRemoved] = ans;
    }
};
